src/app.provider.ts

```diff
@@ -0,0 +1,98 @@
+import { INestApplication, VersioningType } from '@nestjs/common';
+import { NestFactory } from '@nestjs/core';
+import { TestingModule } from '@nestjs/testing/testing-module';
+import * as winston from 'winston';
+import { format } from 'winston';
+
+function configureLogger() {
+  winston.add(
+    new winston.transports.Console({
+      level: 'debug',
+      format: format.combine(format.splat(), format.simple()),
+    }),
+  );
+}
+
+function configureVersioning(app: INestApplication) {
+  app.enableVersioning({
+    type: VersioningType.URI,
+  });
+}
+
+function configureCors(app: INestApplication) {
+  app.enableCors({
+    origin: '*',
+  });
+}
+
+const DEFAULT_CONFIGURATION: ((app: INestApplication) => void)[] = [
+  configureLogger,
+  configureVersioning,
+  configureCors,
+];
+
+/**
+ * The main goal of {@link AppProvider} is to provide
+ * a {@link INestApplication}.
+ *
+ * Extensions of this class should return the application in
+ * {@link getApp}.
+ *
+ * Each provider should have a {@link configuration} which specifies
+ * the steps taken to configure the application
+ */
+export abstract class AppProvider {
+  protected abstract readonly configuration: Array<
+    (app: INestApplication) => void
+  >;
+
+  public async provide(module: any): Promise<INestApplication> {
+    const app = await this.getApp(module);
+    this.configuration.forEach((f) => f(app));
+    return app;
+  }
+
+  protected abstract getApp(module: any): Promise<INestApplication>;
```

Comment:
Nit: can we use the NestJS types instead of `any` here?

Reply:
Same with the other `any`s here.

Reply:
Double checking because as far as I remember the type of the modules on NestJS side is not-strict and it's actually any 👀

Reply:
Here: https://github.com/nestjs/nest/blob/6d17c2a869c803e33600453009ff3bd77e69195b/packages/core/nest-factory.ts#L50